### PR TITLE
Added a basic buildpack test & fixed checker discovery

### PR DESCRIPTION
**Pull request by dorfire**

### CODE REVIEW

File: src/BuildScriptGenerator/Node/NodePlatform.cs [hunk 1]
```diff
@@ -483,10 +488,8 @@ private string GetNpmVersion(dynamic packageJson)
             if (!string.IsNullOrWhiteSpace(npmVersionRange))
             {
                 var supportedNpmVersions = _nodeVersionProvider.SupportedNpmVersions;
-                npmVersion = SemanticVersionResolver.GetMaxSatisfyingVersion(
-                    npmVersionRange,
-                    supportedNpmVersions);
-                if (string.IsNullOrWhiteSpace(npmVersion))
+                npmVersion = SemanticVersionResolver.GetMaxSatisfyingVersion(npmVersionRange, supportedNpmVersions);
+                if (string.IsNullOrEmpty(npmVersion))
```
Comment (gpcastro) [inline src/BuildScriptGenerator/Node/NodePlatform.cs]: why this change?

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs [hunk 2]
```diff
@@ -208,16 +216,24 @@ private static string GetBenvArgs(IDictionary<string, string> benvArgsMap)
             IDictionary<string, string> tools,
             [NotNull] List<ICheckerMessage> checkerMessageSink)
         {
-            var checkers = _checkers.WhereApplicable(tools);
+            var checkers = _checkers.WhereApplicable(tools).ToArray();
```
Comment (gpcastro) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: all of these parameters should be `[NotNull]`

File: src/BuildScriptGenerator/Node/NodeDependencyChecker.cs [hunk 3]
```diff
@@ -19,11 +20,19 @@ public class NpmDependencyChecker : IChecker
...
+        {
+            _logger = logger;
+        }
+
         public IEnumerable<ICheckerMessage> CheckSourceRepo(ISourceRepo repo)
         {
             dynamic packageJson = NodePlatform.GetPackageJsonObject(repo, null);
             if (packageJson == null)
             {
+                _logger.LogDebug("packageJson is null");
```
Comment (gpcastro) [inline src/BuildScriptGenerator/Node/NodeDependencyChecker.cs]: It would be helpful to have more context on this message as we read through the logs, to provide a semantic value rather than pure state. e.g. "Node checker skipping repo since packageJson is null"

File: src/BuildScriptGenerator/Node/NodePlatform.cs [hunk 4]
```diff
@@ -203,12 +203,17 @@ public bool IsEnabledForMultiPlatformBuild(BuildScriptGeneratorContext scriptGen
...
+                string npmVersion = GetNpmVersion(packageJson);
+                _logger.LogDebug("GetNpmVersion returned {npmVersion}", npmVersion);
                 if (!string.IsNullOrEmpty(npmVersion))
                 {
                     toolsToVersion[NodeConstants.NpmToolName] = npmVersion;
                 }
             }
+            else
+            {
+                _logger.LogDebug("packageJson is null");
```
Comment (gpcastro) [inline src/BuildScriptGenerator/Node/NodePlatform.cs]: adding more context would be help, e.g. "Not settings Node tools version since packageJson is null"

File: src/BuildScriptGenerator/Node/NodeVersionChecker.cs [hunk 5]
```diff
@@ -6,26 +6,33 @@
...
-                if (SemanticVersionResolver.CompareVersions(used, NodeScriptGeneratorOptionsSetup.NodeLtsVersion) < 0)
+                return new[]
                 {
-                    return new[]
-                    {
-                        new CheckerMessage($"An outdated version of Node.js was used ({used}). Consider updating. " +
-                                           $"Versions supported in Oryx: {Constants.OryxGitHubUrl}")
-                    };
-                }
+                    new CheckerMessage($"An outdated version of Node.js was detected ({used}). Consider updating.\n" +
```
Comment (gpcastro) [inline src/BuildScriptGenerator/Node/NodeVersionChecker.cs]: we have a resource file on this project to put labels in. We should store those strings there so its easier to review them.

File: tests/Oryx.Integration.Tests/NodeEndToEndTests.cs [hunk 6]
```diff
@@ -252,14 +251,47 @@ public async Task CanBuildAndRun_NodeApp(string nodeVersion)
                 });
         }
 
+        [Fact]
+        public async Task CanBuildAndRun_NodeApp_WithBuildpack()
+        {
+            // Arrange
+            var appName = "webfrontend";
+            var appVolume = CreateAppVolume(appName);
+            var dockerPort = new DockerVolume("/var/run/docker.sock", "/var/run/docker.sock");
```
Comment (kichalla) [inline tests/Oryx.Integration.Tests/NodeEndToEndTests.cs]: Curious, what does this do?

File: tests/Oryx.Tests.Common/EndToEndTestHelper.cs [hunk 7]
```diff
@@ -130,7 +161,7 @@ public static class EndToEndTestHelper
             // Build
             var buildAppResult = dockerCli.Run(new DockerRunArguments
             {
-                ImageId = Settings.BuildImageName,
+                ImageId = buildImage,
```
Comment (kichalla) [inline tests/Oryx.Tests.Common/EndToEndTestHelper.cs]: 👍 I wanted to make this change for a while now

File: tests/Oryx.Tests.Common/DockerVolume.cs [hunk 8]
```diff
@@ -22,6 +22,15 @@ public class DockerVolume
 
         public const string ContainerDirRoot = "/oryxtests";
 
+        // TODO: rename Dir -> Path
+        public DockerVolume(
```
Comment (kichalla) [inline tests/Oryx.Tests.Common/DockerVolume.cs]: Not making this public was deliberate since end users (us who write tests) do not hit into errors, so I am wondering if this change is really required for your use case.

File: src/BuildScriptGeneratorCli/Commands/BuildCommand.cs [hunk 9]
```diff
@@ -140,16 +140,22 @@ internal override int Execute(IServiceProvider serviceProvider, IConsole console
...
                 var scriptGenerator = new BuildScriptGenerator(serviceProvider, console, checkerMessages);
 
+                var generated = scriptGenerator.TryGenerateScript(out scriptContent);
+                stopwatch.AddProperty("generateSucceeded", generated.ToString());
+
                 if (checkerMessages.Count > 0)
                 {
                     var messageFormatter = new DefinitionListFormatter();
-                    checkerMessages.Select(msg => messageFormatter.AddDefinition(msg.Level.ToString(), msg.Content));
+                    checkerMessages.ForEach(msg => messageFormatter.AddDefinition(msg.Level.ToString(), msg.Content));
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/BuildCommand.cs]: Do we need tests for these lines of code?

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs [hunk 10]
```diff
@@ -255,13 +271,14 @@ private static string GetBenvArgs(IDictionary<string, string> benvArgsMap)
...
                 {
-                    _logger.LogDebug("Script generator {scriptGenType} was used", platform.GetType());
+                    _logger.LogDebug("Platform {platformType} was used", platform.GetType());
                     snippets.Add(snippet);
                     platform.SetRequiredTools(context.SourceRepo, targetVersion, toolsToVersion);
                 }
                 else
                 {
-                    _logger.LogDebug("Script generator {scriptGenType} cannot be used", platform.GetType());
+                    _logger.LogWarning("{platformType}.GenerateBashBuildScriptSnippet() returned null",
```
Comment (kichalla) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: Split the arguments to a method into separate lines if you are not on a single line

File: src/BuildScriptGenerator/Node/NodePlatform.cs (hunk 1 above)
Comment (dorfire) [reply src/BuildScriptGenerator/Node/NodePlatform.cs]: because it fits in less than 120 characters

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs (hunk 2 above)
Comment (dorfire) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: correct, but i wanted to be over-explicit and avoid confusion with `IBuildScriptGenerator.GenerateBashScript` which can receive a null checkerMessageSink.

File: src/BuildScriptGenerator/Node/NodeDependencyChecker.cs (hunk 3 above)
Comment (dorfire) [reply src/BuildScriptGenerator/Node/NodeDependencyChecker.cs]: Don't forget this message will have a `LoggerName` costume dimension with it - do you think the extra context is still worth adding?

File: tests/Oryx.Integration.Tests/NodeEndToEndTests.cs (hunk 6 above)
Comment (dorfire) [reply tests/Oryx.Integration.Tests/NodeEndToEndTests.cs]: Allows the `pack` binary to use the Docker daemon

File: tests/Oryx.Tests.Common/DockerVolume.cs (hunk 8 above)
Comment (dorfire) [reply tests/Oryx.Tests.Common/DockerVolume.cs]: Well, I don't really need the whole mirror-copy thing that happens in `DockerVolume.Create`.. so the only other choice is introducing another static initializer (like renaming the current method to `CreateMirror` and then adding a no-op `DockerVolume.Create`)

File: src/BuildScriptGeneratorCli/Commands/BuildCommand.cs (hunk 9 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/BuildCommand.cs]: We can modify an existing end-to-end test to cover them

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs (hunk 10 above)
Comment (dorfire) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: I changed the styling rule file so that this doesn't trigger a warning anymore

File: tests/Oryx.Tests.Common/DockerVolume.cs (hunk 8 above)
Comment (kichalla) [reply tests/Oryx.Tests.Common/DockerVolume.cs]: Yes, i was thinking the same thing...we should rather create another static member to abstract the user from this detail.

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs (hunk 10 above)
Comment (kichalla) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: Some teams (ex: ASP.NET Core) follow this pattern of either having all args on a single line or split individually...we could all agree on one approach and stick to it...

File: src/BuildScriptGenerator/Node/NodeDependencyChecker.cs (hunk 3 above)
Comment (gpcastro) [reply src/BuildScriptGenerator/Node/NodeDependencyChecker.cs]: Yes. It's easier to see what's going on when reading through the files. Besides, this message depends on the existing version of the code to see what that means. If we change this code at some point, we may not be able to understand the right context in which this was presented.
